Assets/Mirror/Runtime/QuaternionReadWrite.cs

```diff
@@ -0,0 +1,348 @@
+        {
+            int index = 0;
+            float current = Mathf.Abs(q.x);
+
+            for (int i = 1; i < 4; i++)
+            {
+                float next = Mathf.Abs(q[i]);
+                if (next > current)
+                {
+                    index = i;
+                    current = next;
+                }
+            }
+
+            return index;
+        }
+
+
+        static Vector3 getSmallerDimensions(int largestIndex, Quaternion value)
+        {
+            float x = value.x;
+            float y = value.y;
+            float z = value.z;
+            float w = value.w;
+
+            float a, b, c;
+            switch (largestIndex)
+            {
+                case 0:
+                    a = y;
+                    b = z;
+                    c = w;
+                    break;
+                case 1:
+                    a = x;
+                    b = z;
+                    c = w;
+                    break;
+                case 2:
+                    a = x;
+                    b = y;
+                    c = w;
+                    break;
+                case 3:
+                    a = x;
+                    b = y;
+                    c = z;
+                    break;
+                default:
+                    throw new System.ArgumentException("LargestIndex did not have value between 0 and 3");
+            }
+            // largest needs to be positive to be calculated by reader 
+            // if largest is negative flip sign of others
+            float largestSign = Mathf.Sign(value[largestIndex]);
+            return new Vector3(a, b, c) * largestSign;
+        }
+        static Quaternion FromSmallerDimensions(uint largestIndex, Vector3 smallest)
+        {
+            float a = smallest.x;
+            float b = smallest.y;
+            float c = smallest.z;
+
+            float x, y, z, w;
+            float largest = Mathf.Sqrt(1 - a * a - b * b - c * c);
+            switch (largestIndex)
+            {
+                case 0:
+                    x = largest;
+                    y = a;
+                    z = b;
+                    w = c;
+                    break;
+                case 1:
+                    x = a;
+                    y = largest;
+                    z = b;
+                    w = c;
+                    break;
+                case 2:
+                    x = a;
+                    y = b;
+                    z = largest;
+                    w = c;
+                    break;
+                case 3:
+                    x = a;
+                    y = b;
+                    z = c;
+                    w = largest;
+                    break;
+                default:
+                    throw new System.ArgumentException("LargestIndex did not have value between 0 and 3");
+            }
+            // 0.999999f is tolerance for Unity's Dot function
+            Debug.Assert(x * x + y * y + z * z + w * w < (2 - 0.999999f), "larger than 1");
```

Comment:
in which case would this assert fire? what does the user have to do wrong?
also, does unity have 0.999999f in a variable somewhere so we can use it and not rely on hardcoding?

Reply:
The assert is there to check the code above is working. I can try replace it with a unit test it that sounds better?